### PR TITLE
[WIP] Build optimizations

### PR DESCRIPTION
**Pull request by divmain**
This PR is probably worth looking at now.  All tests pass, although I do intend to add additional tests.  I also need to pull in latest `canary`; ~~please advise on direction for the question at the end of this PR description, and that'll inform how I pull in latest `canary` commits.~~

Measuring wins was difficult; I was looking for a good React project (or projects) that could be used to test performance changes (both for this PR and others), but without much success.  I put together [fuzzponent](https://github.com/divmain/fuzzponent) for this purpose.  Given a seed and a recurse-depth, it'll generate a large tree of React component JS files, PRNG-style.

Performance has definitely improved with the changes in this PR.  To test, I generated a project of 3021 React components with `fuzzponent -d 2 -s 206`.  If you run that command, you should get the same components.

- No cache (avg): 20.64 seconds
- Cold cache (avg): 17.76 seconds (14% improvement)
- Warm cache (avg): 12.19 seconds (41% improvement)

All times were measured w/ `time next build .`.  I haven't yet investigated dev build times; I don't expect to see significant changes.  Cold cache times improved because server build step improved (client build warmed the cache).

The loader exposes an API for a non-FS-backed cache.  This could be an in-memory cache, Redis, or something else.

<details><summary><i>(no longer relevant)</i></summary>
<p>

There are a _ton_ of commits showing here - that is because I preserved the git commit history for `babel-loader` when pulling it in with:

```
git filter-branch -f --tree-filter '
    if [ -d ./src ]; then
        find . -depth 1 -not -name "src" -not -name ".git" -not -name "LICENSE" -print0 | xargs -0 rm -fr
        mkdir -p packages/next/build/webpack/loaders
        mv src packages/next/build/webpack/loaders/next-babel-loader
        mv packages/next/build/webpack/loaders/next-babel-loader/index.js packages/next/build/webpack/loaders/next-babel-loader/babel-loader.js
        if [ -f ./LICENSE ]; then
            mv LICENSE packages/next/build/webpack/loaders/next-babel-loader/
        fi
    else 
        find . -depth 1 -not -name ".git" -print0 | xargs -0 rm -fr
    fi
' HEAD
```

I can squash all of that if you don't see value in having history for those files.

</p>
</details>

@timneutkens @rauchg

### CODE REVIEW

File: packages/next/package.json [hunk 1]
```diff
@@ -49,13 +49,14 @@
     "ansi-html": "0.0.7",
     "autodll-webpack-plugin": "0.4.2",
     "babel-core": "7.0.0-bridge.0",
-    "babel-loader": "8.0.2",
+    "babel-loader": "github:divmain/babel-loader#master",
```
Comment (timneutkens) [inline packages/next/package.json]: This should be removed right?

File: packages/next/package.json [hunk 2]
```diff
@@ -77,6 +79,7 @@
     "terser-webpack-plugin": "1.1.0",
     "unfetch": "3.0.0",
     "url": "0.11.0",
+    "util.promisify": "1.0.0",
```
Comment (timneutkens) [inline packages/next/package.json]: You can use `import {promisify} from 'util'` since we support Node.js 8+

File: packages/next/package.json [hunk 3]
```diff
@@ -64,6 +65,7 @@
     "launch-editor": "2.2.1",
     "loader-utils": "1.1.0",
     "minimist": "1.2.0",
+    "mkdirp": "0.5.1",
```
Comment (timneutkens) [inline packages/next/package.json]: Either `mkdirp` or `mkdirp-then` can be removed. I'd vote for `mkdirp-then` and just using `promisify` 🤔

**Comment by timneutkens** [comment]
We'll squash the PR, so I guess keeping the history here is fine either way. However I think we're pinging everyone that ever contributed to babel-loader with this as the PR is showing 56 participants

File: packages/next/build/webpack.js [hunk 4]
```diff
@@ -172,7 +176,18 @@ export default async function getBaseWebpackConfig (dir, {dev = false, isServer
...
-      options: {dev, isServer, cwd: dir}
+      options: {
+        dev,
+        isServer,
+        cwd: dir,
+        cacheIdentifier: JSON.stringify({
+          config,
+          'next': NEXT_VERSION,
+          '@babel/core': babelPkg.version,
+          '@babel/loader': babelLoaderPkg.version
```
Comment (timneutkens) [inline packages/next/build/webpack.js]: Guess this line has to be removed since we have babel-loader inside next after this PR.

**Comment by divmain** [comment]
Squashed the commits to remove all the extra participants (sorry about that!).  Will address other feedback later today.  Thanks!

**Comment by vjpr** [comment]
@divmain 

From what I gather, this PR implements a custom babel webpack loader with a file system cache.

Since webpack5 will have built-in persistent caching (https://twitter.com/TheLarkInn/status/1069246795031576576), will we still need this?

I am keen to have an FS-backed cache for next.js asap though, so this looks great!

**Comment by vjpr** [comment]
@divmain Also, would this loader be useful for other webpack projects? 

Maybe its better to create a separate repo for it.

Also, not sure how close it is to being merged and released, but it might be an idea to override the `next-babel-loader` in `next.config.js` and move the code to separate repo, so we can test it.

**Comment by vjpr** [comment]
Just occurred to me that this is will not really improve things more than babel's `cacheDirectory`, it would just allow more caching options such as a redis db that could be shared with a team.

File: packages/next/package.json [hunk 5]
```diff
@@ -69,7 +69,7 @@
     "http-status": "1.0.1",
     "launch-editor": "2.2.1",
     "loader-utils": "1.1.0",
-    "mkdirp-then": "1.2.0",
+    "mkdirp": "0.5.1",
```
Comment (wtgtybhertgeghgtwtg) [inline packages/next/package.json]: I'd go with `make-dir` since it's already in the dependency tree via `find-cache-dir`.

**Review by timneutkens — COMMENTED** [review]
I'm going to close this PR for now, we'll revisit it 🔜

**Pull request closed without merging**